features/graph/provider/application/src/main/java/org/opennms/netmgt/graph/provider/application/ApplicationGraphProvider.java

```diff
@@ -0,0 +1,123 @@
+ * it under the terms of the GNU Affero General Public License as published
+ * by the Free Software Foundation, either version 3 of the License,
+ * or (at your option) any later version.
+ *
+ * OpenNMS(R) is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with OpenNMS(R).  If not, see:
+ *      http://www.gnu.org/licenses/
+ *
+ * For more information contact:
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.netmgt.graph.provider.application;
+
+import java.util.Objects;
+
+import org.opennms.netmgt.dao.api.ApplicationDao;
+import org.opennms.netmgt.events.api.EventListener;
+import org.opennms.netmgt.graph.api.Graph;
+import org.opennms.netmgt.graph.api.info.DefaultGraphInfo;
+import org.opennms.netmgt.graph.api.info.GraphInfo;
+import org.opennms.netmgt.graph.api.service.GraphProvider;
+import org.opennms.netmgt.graph.simple.SimpleEdge;
+import org.opennms.netmgt.graph.simple.SimpleGraph;
+import org.opennms.netmgt.graph.simple.SimpleVertex;
+import org.opennms.netmgt.model.OnmsApplication;
+import org.opennms.netmgt.model.OnmsMonitoredService;
+import org.opennms.netmgt.xml.event.Event;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ApplicationGraphProvider implements GraphProvider, EventListener {
+
+    static final String TOPOLOGY_NAMESPACE = "application";
+
+    private static final Logger LOG = LoggerFactory.getLogger(ApplicationGraphProvider.class);
+
+    private Graph graph;
+
+    private ApplicationDao applicationDao;
+
+    private boolean initialized = false;
+
+    public ApplicationGraphProvider(ApplicationDao applicationDao) {
+        Objects.requireNonNull(applicationDao);
+        this.applicationDao = applicationDao;
+        LOG.debug("Creating a new {} with namespace {}", getClass().getSimpleName(), TOPOLOGY_NAMESPACE);
+    }
+
+    // TODO MVR We may need some kind of caching strategy implementation allowing each provider to deal with reloads individually if they so choose
+    @Override
+    public Graph<?, ?> loadGraph() {
+        // TODO MVR this is not thread safe
+        if (!initialized) {
+            graph = createGraph();
+            initialized = true;
+        }
+        return graph;
+    }
+
+    @Override
+    public GraphInfo<?> getGraphInfo() {
+        final DefaultGraphInfo graphInfo = new DefaultGraphInfo(TOPOLOGY_NAMESPACE, SimpleVertex.class);
+        graphInfo.setLabel("Application Graph");
+        graphInfo.setDescription("This Topology Provider displays the hierarchy of TODO."); // TODO Patrick
```

Comment:
Uhm what does it actually display? Applications with its monitored services?

Reply:
> This Topology Provider displays all defined Applications and their calculated states.
Source https://github.com/opennms/opennms/blob/09a8e5126ee824d177b4f8b322072a111384f729/features/topology-map/plugins/org.opennms.features.topology.plugins.topo.application/src/main/resources/OSGI-INF/blueprint/blueprint.xml#L13-L17

Also refer to its implementation https://github.com/opennms/opennms/blob/47a0b591e9dd04d6b3bef3d3765bac13eb12cf35/features/topology-map/plugins/org.opennms.features.topology.plugins.topo.application/src/main/java/org/opennms/features/topology/plugins/topo/application/ApplicationTopologyProvider.java#L55

Reply:
fixed.